### PR TITLE
Allow configuring "negative" DNS TTL for failed resolutions

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

Both JDK and Netty resolvers allow configuring "negative TTL" for
caching unsuccessful DNS resolutions. By default, it's configured to
0, but there is a Java system property to change it for
`java.net.InetAddress`.

Modifications:

- Add `DnsServiceDiscovererBuilder#ttl(int, int, int, int, int)`
overload that allows configuring negative TTL;
- Recognize `networkaddress.cache.negative.ttl` system property to
adjust the default value for Netty resolver;
- Log the default value at debug level;

Result:

Users can configure DNS caching for unsuccessful results.

### CODE REVIEW

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java [hunk 1]
```diff
@@ -65,20 +67,38 @@ public final class DefaultDnsServiceDiscovererBuilder implements DnsServiceDisco
     private static final int DEFAULT_MIN_TTL_POLL_SECONDS = 10;
     private static final int DEFAULT_MAX_TTL_POLL_SECONDS = (int) TimeUnit.MINUTES.toSeconds(5);
     private static final int DEFAULT_MIN_TTL_CACHE_SECONDS = 0;
-    private static final int DEFAULT_MAX_TTL_CACHE_SECONDS = 0;
+    private static final int DEFAULT_MAX_TTL_CACHE_SECONDS = 30;
+    private static final int DEFAULT_TTL_POLL_JITTER_SECONDS = 4;
+    // System property recognized by JVM: https://docs.oracle.com/javase/8/docs/technotes/guides/net/properties.html
```
Comment (Scottmitch) [inline servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java:72]: nit: consider putting in java doc comment `/** <a href=".."></a> */`

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java [hunk 2]
```diff
@@ -170,11 +212,10 @@ public DefaultDnsServiceDiscovererBuilder ttl(final int minSeconds, final int ma
                     " (expected: 0 <= minCacheSeconds <= minSeconds(" + minSeconds +
                     ") <= maxCacheSeconds <= maxSeconds(" + maxSeconds + "))");
         }
-        this.minTTLSeconds = minSeconds;
-        this.maxTTLSeconds = maxSeconds;
-        this.minTTLCacheSeconds = minCacheSeconds;
-        this.maxTTLCacheSeconds = maxCacheSeconds;
-        return this;
+        if (negativeTTLCacheSeconds < 0) {
```
Comment (Scottmitch) [inline servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java:215]: this is called in a static context and will prevent the app from starting which maybe difficult to understand. did you consider logging an error and choosing a default value if invalid values are specified?

Comment (idelpivnitskiy) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java:202]: Actually, it's a leftover from some intermediate work when I had more properties. With only one property, there is no need to validate inside the static block as we always choose the default, even if we can not parse the value.

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java [hunk 3]
```diff
@@ -68,9 +69,22 @@ public final class DefaultDnsServiceDiscovererBuilder implements DnsServiceDisco
...
+     * original TTL from the server and caches exactly for the specified amount of time instead of just being a max cap.
+     */
+    private static final String NEGATIVE_TTL_CACHE_SECONDS_PROPERTY = "networkaddress.cache.negative.ttl";
+    private static final int DEFAULT_NEGATIVE_TTL_CACHE_SECONDS;
     private static final ServiceDiscovererEvent.Status DEFAULT_MISSING_RECOREDS_STATUS = EXPIRED;
 
     static {
+        final Integer negativeCacheTtlValue = parseProperty(NEGATIVE_TTL_CACHE_SECONDS_PROPERTY);
+        DEFAULT_NEGATIVE_TTL_CACHE_SECONDS = negativeCacheTtlValue == null ? 0 :
+                // A value of -1 indicates "cache forever".
```
Comment (bryce-anderson) [inline servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java:85]: The way it's coded it's a value less than 0 indicates "cache forever".

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java [hunk 4]
```diff
@@ -339,11 +369,26 @@ private static DnsClientFilterFactory appendFilter(@Nullable final DnsClientFilt
...
+        final String propertyValue = getProperty(propertyName);
+        if (propertyValue == null) {
+            return null;
+        }
+        try {
+            return Integer.valueOf(propertyValue);
+        } catch (NumberFormatException e) {
+            // Ignore
+            return null;
+        }
```
Comment (bryce-anderson) [inline servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java:392]: I agree with @Scottmitch that if this happens we should log it as an error.

**Review by bryce-anderson — APPROVED** [review]

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java (hunk 3 above)
Comment (idelpivnitskiy) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java]: Updated the comment to clarify

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DnsServiceDiscovererBuilder.java [hunk 5]
```diff
@@ -72,9 +73,45 @@ public interface DnsServiceDiscovererBuilder {
...
+     * than or equal to {@code minCacheSeconds}, and less than or equal to {@code maxSeconds}.
+     * @param negativeCacheSeconds The amount of time an unsuccessful (failed) result will be cached locally (in
+     * seconds), must be greater than or equal to {@code 0}. If other overloads are used, the default value will
+     * recognize the standard Java system property {@code networkaddress.cache.negative.ttl},
+     * like {@link java.net.InetAddress} does.
+     * @return {@code this}.
+     * @see #ttl(int, int)
+     * @see #ttl(int, int, int, int)
+     */
+    default DnsServiceDiscovererBuilder ttl(int minSeconds, int maxSeconds, int minCacheSeconds, int maxCacheSeconds,
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DnsServiceDiscovererBuilder.java:109]: I'm agreed that 5 look like at the limit. Taking into account that:
- it's a very rare use-case that someone needs this overload
- we already have 2 other overloads with the same pattern
- proposed approach requires designing more public API classes and deprecating existing overloads
I would propose sticking to the pattern and shipping as-is.
In general, it's rare that users change any settings on this builder.

**Review by daschl — APPROVED** [review]

**Pull request merged**